File: android/src/main/java/com/risemaxi/graft/classes/ChannelDocument.java
package com.risemaxi.graft.classes;

import androidx.annotation.NonNull;
import java.util.ArrayList;
import java.util.List;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/**
 * The list of releases published to a channel. It is served unsigned and edge-cached, so it is only
 * ever used to choose a manifest to fetch.
 */
public class ChannelDocument {

    private static final int SUPPORTED_SCHEMA = 1;

    private final boolean killSwitch;

    @NonNull
    private final List<ChannelRelease> releases = new ArrayList<>();

    public ChannelDocument(@NonNull JSONObject json) throws JSONException {
        int schema = json.getInt("schema");
        if (schema != SUPPORTED_SCHEMA) {
            throw new JSONException("Unsupported channel document schema: " + schema);
        }
        this.killSwitch = json.optBoolean("killSwitch", false);
        JSONArray releasesJson = json.getJSONArray("releases");
        for (int i = 0; i < releasesJson.length(); i++) {
            releases.add(new ChannelRelease(releasesJson.getJSONObject(i)));
        }
    }

    public boolean isKillSwitchEnabled() {
        return killSwitch;
    }

    @NonNull
    public List<ChannelRelease> getReleases() {
        return releases;
    }
}

File: ios/Sources/RisemaxiGraft/Classes/ChannelDocument.swift
import Foundation

/// The list of releases published to a channel. It is served unsigned and edge-cached, so it is only
/// ever used to choose a manifest to fetch.
public struct ChannelDocument: Decodable {
    private static let supportedSchema = 1

    let killSwitch: Bool
    let releases: [ChannelRelease]

    private enum CodingKeys: String, CodingKey {
        case schema, killSwitch, releases
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let schema = try container.decode(Int.self, forKey: .schema)
        guard schema == ChannelDocument.supportedSchema else {
            throw DecodingError.dataCorruptedError(forKey: .schema, in: container, debugDescription: "Unsupported channel document schema: \(schema)")
        }
        killSwitch = try container.decodeIfPresent(Bool.self, forKey: .killSwitch) ?? false
        releases = try container.decode([ChannelRelease].self, forKey: .releases)
    }
}
